File: code/packages/rust/grammar-tools/src/bin/grammar-tools.rs
//! # grammar-tools CLI — validate `.tokens` and `.grammar` files from the command line.
//!
//! This binary exposes the validation functions from the `grammar_tools` library
//! as a command-line tool. It is the Rust counterpart of the Python
//! `python -m grammar_tools` CLI and produces identical output so that CI
//! scripts can use either implementation interchangeably.
//!
//! # Why a CLI binary?
//!
//! The library functions (`validate_token_grammar`, `validate_parser_grammar`,
//! `cross_validate`) exist as pure Rust functions, but grammar authors need a
//! quick way to check files from the terminal — without writing any Rust code.
//! Think of this like a compiler's `-fsyntax-only` flag: parse and validate,
//! report what is wrong, exit non-zero on failure.
//!
//! # Usage
//!
//! ```text
//! grammar-tools validate <file.tokens> <file.grammar>
//! grammar-tools validate-tokens <file.tokens>
//! grammar-tools validate-grammar <file.grammar>
//! grammar-tools --help
//! ```
//!
//! # Exit codes
//!
//! | Code | Meaning                                      |
//! |------|----------------------------------------------|
//! |  0   | All checks passed                            |
//! |  1   | One or more validation errors found          |
//! |  2   | Wrong number of arguments (usage error)      |
//!
//! # Output format (consistent with Python implementation)
//!
//! Success:
//! ```text
//! Validating lattice.tokens ... OK (N tokens, M skip, K error)
//! Validating lattice.grammar ... OK (P rules)
//! Cross-validating ... OK
//!
//! All checks passed.
//! ```
//!
//! Failure:
//! ```text
//! Validating broken.tokens ... 2 error(s)
//!   Line 5: Duplicate token name 'IDENT' ...
//! Found 4 error(s). Fix them and try again.
//! ```

use std::collections::HashSet;
use std::path::Path;
use std::process;

use grammar_tools::cross_validator::cross_validate;
use grammar_tools::parser_grammar::{parse_parser_grammar, validate_parser_grammar};
use grammar_tools::token_grammar::{parse_token_grammar, token_names, validate_token_grammar};
use grammar_tools::compiler::{compile_tokens_to_rust, compile_parser_to_rust};

// ===========================================================================
// Count errors (vs warnings)
// ===========================================================================

/// Count how many issues are actual errors (not warnings).
///
/// Issues starting with `"Warning"` are informational and do not cause the
/// tool to fail. Everything else counts as a real error.
///
/// This mirrors `_count_errors()` in the Python implementation.
fn count_errors(issues: &[String]) -> usize {
    issues.iter().filter(|i| !i.starts_with("Warning")).count()
}

/// Print a list of issues with two-space indentation.
fn print_issues(issues: &[String]) {
    for issue in issues {
        println!("  {}", issue);
    }
}

// ===========================================================================
// validate command — validate a (.tokens, .grammar) pair
// ===========================================================================

/// Validate a `.tokens` file and a `.grammar` file together.
///
/// This is the core of the `validate` subcommand:
/// 1. Parse the `.tokens` file and run `validate_token_grammar`.
/// 2. Parse the `.grammar` file and run `validate_parser_grammar`.
/// 3. Cross-validate the two with `cross_validate`.
///
/// Returns 0 if all checks pass, 1 if any errors are found.
fn validate_command(tokens_path: &str, grammar_path: &str) -> i32 {
    let mut total_errors: usize = 0;

    // -----------------------------------------------------------------------
    // Step 1: Parse and validate the .tokens file.
    // -----------------------------------------------------------------------
    let tokens_filename = Path::new(tokens_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(tokens_path);

    print!("Validating {} ... ", tokens_filename);

    let tokens_source = match std::fs::read_to_string(tokens_path) {
        Ok(s) => s,
        Err(e) => {
            println!("ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    let token_grammar = match parse_token_grammar(&tokens_source) {
        Ok(g) => g,
        Err(e) => {
            println!("PARSE ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    let token_issues = validate_token_grammar(&token_grammar);
    let token_errors = count_errors(&token_issues);
    let n_tokens = token_grammar.definitions.len();
    let n_skip = token_grammar.skip_definitions.len();
    let n_error_defs = token_grammar.error_definitions.len();

    if token_errors > 0 {
        println!("{} error(s)", token_errors);
        print_issues(&token_issues);
        total_errors += token_errors;
    } else {
        // Build the "OK (N tokens, M skip, K error)" summary line.
        // Only include non-zero counts to avoid noise (mirrors Python).
        let mut parts = vec![format!("{} tokens", n_tokens)];
        if n_skip > 0 {
            parts.push(format!("{} skip", n_skip));
        }
        if n_error_defs > 0 {
            parts.push(format!("{} error", n_error_defs));
        }
        println!("OK ({})", parts.join(", "));
    }

    // -----------------------------------------------------------------------
    // Step 2: Parse and validate the .grammar file.
    // -----------------------------------------------------------------------
    let grammar_filename = Path::new(grammar_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(grammar_path);

    print!("Validating {} ... ", grammar_filename);

    let grammar_source = match std::fs::read_to_string(grammar_path) {
        Ok(s) => s,
        Err(e) => {
            println!("ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    let parser_grammar = match parse_parser_grammar(&grammar_source) {
        Ok(g) => g,
        Err(e) => {
            println!("PARSE ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    // Pass the token names so undefined token references are caught.
    let tnames = token_names(&token_grammar);
    let parser_issues = validate_parser_grammar(&parser_grammar, Some(&tnames));
    let parser_errors = count_errors(&parser_issues);
    let n_rules = parser_grammar.rules.len();

    if parser_errors > 0 {
        println!("{} error(s)", parser_errors);
        print_issues(&parser_issues);
        total_errors += parser_errors;
    } else {
        println!("OK ({} rules)", n_rules);
    }

    // -----------------------------------------------------------------------
    // Step 3: Cross-validate the two files together.
    // -----------------------------------------------------------------------
    print!("Cross-validating ... ");

    let cross_issues = cross_validate(&token_grammar, &parser_grammar);
    let cross_errors = count_errors(&cross_issues);
    let cross_warnings = cross_issues.len() - cross_errors;

    if cross_errors > 0 {
        println!("{} error(s)", cross_errors);
        print_issues(&cross_issues);
        total_errors += cross_errors;
    } else if cross_warnings > 0 {
        println!("OK ({} warning(s))", cross_warnings);
        print_issues(&cross_issues);
    } else {
        println!("OK");
    }

    // -----------------------------------------------------------------------
    // Summary
    // -----------------------------------------------------------------------
    if total_errors > 0 {
        println!("\nFound {} error(s). Fix them and try again.", total_errors);
        1
    } else {
        println!("\nAll checks passed.");
        0
    }
}

// ===========================================================================
// validate-tokens command — validate only a .tokens file
// ===========================================================================

/// Validate just a `.tokens` file (no grammar file needed).
///
/// Returns 0 if all checks pass, 1 if any errors are found.
fn validate_tokens_only(tokens_path: &str) -> i32 {
    let tokens_filename = Path::new(tokens_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(tokens_path);

    print!("Validating {} ... ", tokens_filename);

    let tokens_source = match std::fs::read_to_string(tokens_path) {
        Ok(s) => s,
        Err(e) => {
            println!("ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    let token_grammar = match parse_token_grammar(&tokens_source) {
        Ok(g) => g,
        Err(e) => {
            println!("PARSE ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    let issues = validate_token_grammar(&token_grammar);
    let errors = count_errors(&issues);
    let n_tokens = token_grammar.definitions.len();

    if errors > 0 {
        println!("{} error(s)", errors);
        print_issues(&issues);
        println!("\nFound {} error(s). Fix them and try again.", errors);
        1
    } else {
        println!("OK ({} tokens)", n_tokens);
        println!("\nAll checks passed.");
        0
    }
}

// ===========================================================================
// validate-grammar command — validate only a .grammar file
// ===========================================================================

/// Validate just a `.grammar` file (no tokens file needed).
///
/// Because we have no `.tokens` file, we can only check rule-level issues
/// (undefined rule references, duplicate rules, naming conventions) — not
/// undefined token references.
///
/// Returns 0 if all checks pass, 1 if any errors are found.
fn validate_grammar_only(grammar_path: &str) -> i32 {
    let grammar_filename = Path::new(grammar_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(grammar_path);

    print!("Validating {} ... ", grammar_filename);

    let grammar_source = match std::fs::read_to_string(grammar_path) {
        Ok(s) => s,
        Err(e) => {
            println!("ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    let parser_grammar = match parse_parser_grammar(&grammar_source) {
        Ok(g) => g,
        Err(e) => {
            println!("PARSE ERROR");
            println!("  {}", e);
            return 1;
        }
    };

    // Without a tokens file, pass None so token-reference checks are skipped.
    let issues = validate_parser_grammar(&parser_grammar, None::<&HashSet<String>>);
    let errors = count_errors(&issues);
    let n_rules = parser_grammar.rules.len();

    if errors > 0 {
        println!("{} error(s)", errors);
        print_issues(&issues);
        println!("\nFound {} error(s). Fix them and try again.", errors);
        1
    } else {
        println!("OK ({} rules)", n_rules);
        println!("\nAll checks passed.");
        0
    }
}

// ===========================================================================
// Help text
// ===========================================================================

/// Print usage information to stdout.
fn print_usage() {
    println!("Usage: grammar-tools <command> [args...]");
    println!();
    println!("Commands:");
    println!("  validate <file.tokens> <file.grammar>  Validate a token/grammar pair");
    println!("  validate-tokens <file.tokens>           Validate just a .tokens file");
    println!("  validate-grammar <file.grammar>         Validate just a .grammar file");
    println!("  compile-tokens <file.tokens> <export_name> Compile just a .tokens file to rust");
    println!("  compile-grammar <file.grammar> <export_name> Compile just a .grammar file to rust");
    println!();
    println!("Examples:");
    println!("  grammar-tools validate css.tokens css.grammar");
    println!("  grammar-tools validate-tokens css.tokens");
    println!("  grammar-tools validate-grammar css.grammar");
    println!("  grammar-tools compile-tokens json.tokens json_tokens");
    println!("  grammar-tools compile-grammar json.grammar json_grammar");
    println!();
    println!("Exit codes:");
    println!("  0   All checks passed");
    println!("  1   One or more validation errors found");
    println!("  2   Wrong number of arguments (usage error)");
}

// ===========================================================================
// main
// ===========================================================================

fn main() {
    let args: Vec<String> = std::env::args().collect();
    // args[0] is the binary name; args[1..] are the user-supplied arguments.
    let user_args = &args[1..];

    let exit_code = if user_args.is_empty()
        || user_args[0] == "--help"
        || user_args[0] == "-h"
        || user_args[0] == "help"
    {
        print_usage();
        0
    } else {
        let command = &user_args[0];
        let rest = &user_args[1..];

        match command.as_str() {
            "validate" => {
                if rest.len() != 2 {
                    eprintln!(
                        "Error: 'validate' requires two arguments: <tokens> <grammar>"
                    );
                    eprintln!();
                    print_usage();
                    2
                } else {
                    validate_command(&rest[0], &rest[1])
                }
            }

            "validate-tokens" => {
                if rest.len() != 1 {
                    eprintln!(
                        "Error: 'validate-tokens' requires one argument: <tokens>"
                    );
                    eprintln!();
                    print_usage();
                    2
                } else {
                    validate_tokens_only(&rest[0])
                }
            }

            "validate-grammar" => {
                if rest.len() != 1 {
                    eprintln!(
                        "Error: 'validate-grammar' requires one argument: <grammar>"
                    );
                    eprintln!();
                    print_usage();
                    2
                } else {
                    validate_grammar_only(&rest[0])
                }
            }

            "compile-tokens" => {
                if rest.len() != 2 {
                    eprintln!(
                        "Error: 'compile-tokens' requires two arguments: <tokens> <export_name>"
                    );
                    eprintln!();
                    print_usage();
                    2
                } else {
                    match std::fs::read_to_string(&rest[0]) {
                        Ok(s) => match parse_token_grammar(&s) {
                            Ok(g) => {
                                let issues = validate_token_grammar(&g);
                                if count_errors(&issues) > 0 {
                                    eprintln!("Error: Cannot compile invalid grammar file.");
                                    print_issues(&issues);
                                    1
                                } else {
                                    print!("{}", compile_tokens_to_rust(&g, &rest[1]));
                                    0
                                }
                            }
                            Err(e) => {
                                eprintln!("PARSE ERROR\n  {}", e);
                                1
                            }
                        },
                        Err(e) => {
                            eprintln!("ERROR\n  {}", e);
                            1
                        }
                    }
                }
            }

            "compile-grammar" => {
                if rest.len() != 2 {
                    eprintln!(
                        "Error: 'compile-grammar' requires two arguments: <grammar> <export_name>"
                    );
                    eprintln!();
                    print_usage();
                    2
                } else {
                    match std::fs::read_to_string(&rest[0]) {
                        Ok(s) => match parse_parser_grammar(&s) {
                            Ok(g) => {
                                let issues = validate_parser_grammar(&g, None::<&HashSet<String>>);
                                if count_errors(&issues) > 0 {
                                    eprintln!("Error: Cannot compile invalid grammar file.");
                                    print_issues(&issues);
                                    1
                                } else {
                                    print!("{}", compile_parser_to_rust(&g, &rest[1]));
                                    0
                                }
                            }
                            Err(e) => {
                                eprintln!("PARSE ERROR\n  {}", e);
                                1
                            }
                        },
                        Err(e) => {
                            eprintln!("ERROR\n  {}", e);
                            1
                        }
                    }
                }
            }

            other => {
                eprintln!("Error: Unknown command '{}'", other);
                eprintln!();
                print_usage();
                2
            }
        }
    };

    process::exit(exit_code);
}

File: code/programs/perl/grammar-tools/grammar-tools.pl
#!/usr/bin/env perl
# grammar-tools — validate and compile .tokens and .grammar files.
#
# This program wraps CodingAdventures::GrammarTools and
# CodingAdventures::GrammarTools::Compiler behind a small CLI. It is the
# Perl counterpart of the Ruby/Go/Rust/TypeScript/Elixir/Lua/Python
# grammar-tools CLIs in this monorepo — same commands, same exit codes.
#
# Usage
# -----
#
#     grammar-tools.pl validate <file.tokens> <file.grammar>
#     grammar-tools.pl validate-tokens <file.tokens>
#     grammar-tools.pl validate-grammar <file.grammar>
#     grammar-tools.pl compile-tokens <file.tokens> [-o <output.pm>]
#     grammar-tools.pl compile-grammar <file.grammar> [-o <output.pm>]
#
# Exit codes
# ----------
#
#   0  All checks passed / compilation succeeded.
#   1  One or more validation errors found / compile error.
#   2  Usage error (wrong number of arguments, unknown command).
#
# Compile commands
# -----------------
#
# The compile commands convert `.tokens`/`.grammar` files into Perl source
# code that embeds the grammar as native data structures, eliminating
# runtime file I/O and parsing in downstream packages. Unlike the Ruby/Go/
# Rust/TypeScript/Elixir ports, these commands do not run a validation
# step before compiling (matching the Lua port) — there is no `--force`
# flag because there is nothing to force past.

use strict;
use warnings;
use FindBin;
use File::Spec;

# ----------------------------------------------------------------------------
# Resolve the repo root so we can find code/packages/perl/grammar-tools/lib.
# Walk up from this file's directory until we find code/specs/grammar-tools.json.
# ----------------------------------------------------------------------------
sub find_root {
    my $dir = File::Spec->rel2abs($FindBin::Bin);
    for (1 .. 20) {
        my $marker = File::Spec->catfile($dir, 'code', 'specs', 'grammar-tools.json');
        return $dir if -e $marker;
        my $parent = File::Spec->catdir($dir, File::Spec->updir);
        my $abs_parent = File::Spec->rel2abs($parent);
        last if $abs_parent eq $dir;
        $dir = $abs_parent;
    }
    return File::Spec->rel2abs($FindBin::Bin);
}

my $ROOT = find_root();
use lib;
lib->import(File::Spec->catdir($ROOT, 'code', 'packages', 'perl', 'grammar-tools', 'lib'));

require CodingAdventures::GrammarTools;
require CodingAdventures::GrammarTools::Compiler;

# ----------------------------------------------------------------------------
# Helpers
# ----------------------------------------------------------------------------

sub count_errors {
    my ($issues) = @_;
    my $n = 0;
    for my $issue (@$issues) {
        $n++ unless $issue =~ /^Warning:/;
    }
    return $n;
}

sub print_issues {
    my ($issues, $indent) = @_;
    $indent //= '  ';
    print STDERR "$indent$_\n" for @$issues;
}

sub basename {
    my ($path) = @_;
    my (undef, undef, $file) = File::Spec->splitpath($path);
    return $file;
}

sub read_file {
    my ($path) = @_;
    open my $fh, '<', $path or return (undef, "$!");
    local $/;
    my $content = <$fh>;
    close $fh;
    return ($content, undef);
}

# ----------------------------------------------------------------------------
# validate_command($tokens_path, $grammar_path)
# ----------------------------------------------------------------------------
sub validate_command {
    my ($tokens_path, $grammar_path) = @_;
    my $total_issues = 0;

    unless (-e $tokens_path) {
        print STDERR "Error: File not found: $tokens_path\n";
        exit 1;
    }
    print "Validating " . basename($tokens_path) . " ... ";

    my ($tsrc, $terr) = read_file($tokens_path);
    my ($token_grammar, $terr2) = CodingAdventures::GrammarTools->parse_token_grammar($tsrc);
    unless ($token_grammar) {
        print "PARSE ERROR\n";
        print "  $terr2\n";
        exit 1;
    }
    my $token_issues = CodingAdventures::GrammarTools->validate_token_grammar($token_grammar);
    my $token_errors = count_errors($token_issues);
    if ($token_errors > 0) {
        print "$token_errors error(s)\n";
        print_issues($token_issues);
        $total_issues += $token_errors;
    }
    else {
        my $n_tokens = scalar @{ $token_grammar->definitions };
        print "OK ($n_tokens tokens)\n";
    }

    unless (-e $grammar_path) {
        print STDERR "Error: File not found: $grammar_path\n";
        exit 1;
    }
    print "Validating " . basename($grammar_path) . " ... ";

    my ($gsrc, $gerr) = read_file($grammar_path);
    my ($parser_grammar, $gerr2) = CodingAdventures::GrammarTools->parse_parser_grammar($gsrc);
    unless ($parser_grammar) {
        print "PARSE ERROR\n";
        print "  $gerr2\n";
        exit 1;
    }
    my $token_names = $token_grammar->token_names;
    my $parser_issues = CodingAdventures::GrammarTools->validate_parser_grammar($parser_grammar, $token_names);
    my $parser_errors = count_errors($parser_issues);
    if ($parser_errors > 0) {
        print "$parser_errors error(s)\n";
        print_issues($parser_issues);
        $total_issues += $parser_errors;
    }
    else {
        my $n_rules = scalar @{ $parser_grammar->{rules} };
        print "OK ($n_rules rules)\n";
    }

    print "\n";
    if ($total_issues > 0) {
        print "Found $total_issues error(s). Fix them and try again.\n";
        return 1;
    }
    print "All checks passed.\n";
    return 0;
}

sub validate_tokens_only {
    my ($tokens_path) = @_;
    unless (-e $tokens_path) {
        print STDERR "Error: File not found: $tokens_path\n";
        exit 1;
    }
    print "Validating " . basename($tokens_path) . " ... ";
    my ($src, $err) = read_file($tokens_path);
    my ($grammar, $perr) = CodingAdventures::GrammarTools->parse_token_grammar($src);
    unless ($grammar) {
        print "PARSE ERROR\n";
        print "  $perr\n";
        return 1;
    }
    my $issues = CodingAdventures::GrammarTools->validate_token_grammar($grammar);
    my $errors = count_errors($issues);
    if ($errors > 0) {
        print "$errors error(s)\n";
        print_issues($issues);
        print "\nFound $errors error(s). Fix them and try again.\n";
        return 1;
    }
    my $n = scalar @{ $grammar->definitions };
    print "OK ($n tokens)\n\nAll checks passed.\n";
    return 0;
}

sub validate_grammar_only {
    my ($grammar_path) = @_;
    unless (-e $grammar_path) {
        print STDERR "Error: File not found: $grammar_path\n";
        exit 1;
    }
    print "Validating " . basename($grammar_path) . " ... ";
    my ($src, $err) = read_file($grammar_path);
    my ($grammar, $perr) = CodingAdventures::GrammarTools->parse_parser_grammar($src);
    unless ($grammar) {
        print "PARSE ERROR\n";
        print "  $perr\n";
        return 1;
    }
    my $issues = CodingAdventures::GrammarTools->validate_parser_grammar($grammar);
    my $errors = count_errors($issues);
    if ($errors > 0) {
        print "$errors error(s)\n";
        print_issues($issues);
        print "\nFound $errors error(s). Fix them and try again.\n";
        return 1;
    }
    my $n = scalar @{ $grammar->{rules} };
    print "OK ($n rules)\n\nAll checks passed.\n";
    return 0;
}

# ----------------------------------------------------------------------------
# compile_tokens_command($tokens_path, $output_path)
# ----------------------------------------------------------------------------
sub compile_tokens_command {
    my ($tokens_path, $output_path) = @_;
    unless (-e $tokens_path) {
        print STDERR "Error: File not found: $tokens_path\n";
        return 1;
    }
    print STDERR "Compiling " . basename($tokens_path) . " ... ";
    my ($src, $err) = read_file($tokens_path);
    my ($grammar, $perr) = CodingAdventures::GrammarTools->parse_token_grammar($src);
    unless ($grammar) {
        print STDERR "PARSE ERROR\n";
        print STDERR "  $perr\n";
        return 1;
    }
    my $code = CodingAdventures::GrammarTools::Compiler::compile_token_grammar($grammar, basename($tokens_path));
    if ($output_path) {
        open my $out, '>', $output_path or do {
            print STDERR "Error: cannot write '$output_path': $!\n";
            return 1;
        };
        print $out $code;
        close $out;
        print STDERR "OK -> $output_path\n";
    }
    else {
        print STDERR "OK\n";
        print $code;
    }
    return 0;
}

sub compile_grammar_command {
    my ($grammar_path, $output_path) = @_;
    unless (-e $grammar_path) {
        print STDERR "Error: File not found: $grammar_path\n";
        return 1;
    }
    print STDERR "Compiling " . basename($grammar_path) . " ... ";
    my ($src, $err) = read_file($grammar_path);
    my ($grammar, $perr) = CodingAdventures::GrammarTools->parse_parser_grammar($src);
    unless ($grammar) {
        print STDERR "PARSE ERROR\n";
        print STDERR "  $perr\n";
        return 1;
    }
    my $code = CodingAdventures::GrammarTools::Compiler::compile_parser_grammar($grammar, basename($grammar_path));
    if ($output_path) {
        open my $out, '>', $output_path or do {
            print STDERR "Error: cannot write '$output_path': $!\n";
            return 1;
        };
        print $out $code;
        close $out;
        print STDERR "OK -> $output_path\n";
    }
    else {
        print STDERR "OK\n";
        print $code;
    }
    return 0;
}

# ----------------------------------------------------------------------------
# dispatch($command, \@files, $output_path)
# ----------------------------------------------------------------------------
sub dispatch {
    my ($command, $files, $output_path) = @_;

    if ($command eq 'validate') {
        if (@$files != 2) {
            print STDERR "Error: 'validate' requires two arguments: <tokens> <grammar>\n";
            return 2;
        }
        return validate_command($files->[0], $files->[1]);
    }
    if ($command eq 'validate-tokens') {
        if (@$files != 1) {
            print STDERR "Error: 'validate-tokens' requires one argument: <tokens>\n";
            return 2;
        }
        return validate_tokens_only($files->[0]);
    }
    if ($command eq 'validate-grammar') {
        if (@$files != 1) {
            print STDERR "Error: 'validate-grammar' requires one argument: <grammar>\n";
            return 2;
        }
        return validate_grammar_only($files->[0]);
    }
    if ($command eq 'compile-tokens') {
        if (@$files != 1) {
            print STDERR "Error: 'compile-tokens' requires one argument: <tokens>\n";
            return 2;
        }
        return compile_tokens_command($files->[0], $output_path);
    }
    if ($command eq 'compile-grammar') {
        if (@$files != 1) {
            print STDERR "Error: 'compile-grammar' requires one argument: <grammar>\n";
            return 2;
        }
        return compile_grammar_command($files->[0], $output_path);
    }

    print STDERR "Error: unknown command '$command'\n";
    return 2;
}

# ----------------------------------------------------------------------------
# main
# ----------------------------------------------------------------------------
sub main {
    my @args = @ARGV;
    unless (@args) {
        print STDERR "Usage: grammar-tools.pl <command> [args...]\n";
        return 2;
    }

    my $command = shift @args;
    my $output_path;
    my @files;
    while (@args) {
        my $arg = shift @args;
        if ($arg eq '-o' || $arg eq '--output') {
            $output_path = shift @args;
        }
        else {
            push @files, $arg;
        }
    }

    return dispatch($command, \@files, $output_path);
}

# `unless caller` — only run when executed directly, not when `require`d
# from a test file. Lets t/01-cli.t require this script and call main(),
# dispatch(), etc. as plain functions without an exit() killing the test
# process.
exit(main()) unless caller;

1;
